src/daft-recordbatch/src/ops/window.rs

```diff
@@ -89,8 +89,108 @@ impl RecordBatch {
         self.union(&window_result)
     }
 
+    #[allow(clippy::too_many_arguments)]
+    pub fn window_agg_dynamic_frame(
+        &self,
+        name: String,
+        agg_expr: &AggExpr,
+        min_periods: i64,
+        dtype: &DataType,
+        frame: &WindowFrame,
+    ) -> DaftResult<Self> {
+        if matches!(frame.frame_type, WindowFrameType::Range) {
+            return Err(DaftError::ValueError(
+                "RANGE frame type is not supported yet for window_agg_dynamic_frame".into(),
+            ));
+        }
+
+        let total_rows = self.len();
+
+        // Convert WindowBoundary to Option<i64>
+        let start_boundary = match &frame.start {
+            WindowBoundary::UnboundedPreceding() => None,
+            WindowBoundary::Offset(offset) => Some(*offset),
+            WindowBoundary::UnboundedFollowing() => {
+                return Err(DaftError::ValueError(
+                    "UNBOUNDED FOLLOWING is not valid as a starting frame boundary".into(),
+                ));
+            }
+        };
+
+        let end_boundary = match &frame.end {
+            WindowBoundary::UnboundedFollowing() => None,
+            WindowBoundary::Offset(offset) => Some(*offset),
+            WindowBoundary::UnboundedPreceding() => {
+                return Err(DaftError::ValueError(
+                    "UNBOUNDED PRECEDING is not valid as an ending frame boundary".into(),
+                ));
+            }
+        };
+
+        self.window_agg_rows_between(
+            agg_expr,
+            &name,
+            dtype,
+            start_boundary,
+            end_boundary,
+            min_periods,
+            total_rows,
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn window_agg_rows_between(
+        &self,
+        agg_expr: &AggExpr,
+        name: &str,
+        dtype: &DataType,
+        start_boundary: Option<i64>,
+        end_boundary: Option<i64>,
+        min_periods: i64,
+        total_rows: usize,
+    ) -> DaftResult<Self> {
+        let null_series = Series::full_null(name, dtype, 1);
+
+        // Use the non-optimized implementation (recalculate for each row)
+        let mut result_series: Vec<Series> = Vec::with_capacity(total_rows);
+
+        for row_idx in 0..total_rows {
+            // Calculate frame bounds for this row
+            let frame_start = match start_boundary {
+                None => 0, // Unbounded preceding
+                Some(offset) => (row_idx as i64 + offset).max(0).min(total_rows as i64) as usize,
```

Comment:
Suggestion: use saturating_add_signed here, which can replace the various casts as well as .max(0)

```suggestion
                Some(offset) => row_idx.saturating_add_signed(offset as isize).min(total_rows),
```